eip-0028.md

```diff
@@ -0,0 +1,130 @@
+However, sending token around is not always desirable. Especially for valuable tokens, users might
+not want to send it away, and doing two transactions (one to send it to the depositary address, one
+to refund it back) costs both time and transaction fees.
+
+To overcome this, ErgoAuth proposes a way to authenticate users trustless to have access to certain
+addresses storing a box. The protocol is trustless in both ways: The users don't need to trust
+the dApp, because the dApp does not get access to funds or secrets. The dApp don't need to trust
+the users or the wallet app, because it can validate the authentication keys.
+
+## ErgoAuth authentication protocol
+
+An authentication with ErgoAuth is driven by a dApp that needs to authenticate a user.
+
+1) The user enters the necessary information in the dApp's UI for the dApp to know if authentication
+is necessary. For example, users might enter their P2PK address (or, instead of manually entering,
+use ErgoPay to send the address to the dApp automatically).
+
+2) The dApp determines that authenticating the user is needed. For this, the dApp prepares a unique
+message that the wallet app should sign with a user's private key, and an Ergo address that the user
+needs to authenticate for. This might be a P2PK address, but in general P2S addresses can be also
+used for authentication.
+
+3) The dApp presents an ErgoAuth link for the user to click and open the wallet app and a QR code
+for mobile users to scan from within the wallet app.
+
+4) The wallet application parses the QR code/link data and obtains a
+`ErgoAuthRequestUrl` to fetch the actual `ErgoAuthRequest` data from
+(see [Data Formats](#data-formats) section).
+
+5) When `ErgoAuthRequest` is obtained, the wallet presents a screen showing that a dApp wants to
+authenticate the user, and the address the request is for. The wallet app should also inform the
+user that no funds or moved and no secrets will leave the device.
+
+6) When the user agrees, the wallet app signs the obtained message from ErgoAuthRequest and sends
+the signed message to the ErgoAuthRequest's replyToUrl.
+
+7) The dApp validates the signed message. When successful, it can proceed with its flow.
+
+## Data Formats
+
+Wallet apps should be able to initiate ErgoAuth both by using URI schemes
+(clickable links) or QR codes.
+
+`ergoauth://<URL>`
+
+An URL is provided without the https prefix. http communication is not allowed except for IP addresses
+(in order to test within a local network).
+
+Examples:
+* `ergoauth://sigmavalley.io/auth/2021-16b8-66c4-b800-6e52-8ce4` will make the wallet app request
+`https://sigmausd.io/auth/2021-16b8-66c4-b800-6e52-8ce4`
+* `ergoauth://192.168.0.1/auth` will make the wallet app request
+`http://192.168.0.1/auth`
+
+#### Response body: ErgoAuthRequest
+
+The wallet application should request URL and obtain the following data (json format)
+
+```
+ErgoAuthRequest:
+  - signingMessage: String
+  - address: String
```

Comment:
this should not be Address, as in general Address may require evaluation. This instead should be serialized SigmaBoolean.

Reply:
I think in most of the cases, a dApp wants to authenticate a holder of a certain box or owner of a P2PK address. So to keep it simple for the dApps, my plan was to use the contract address here from which the ErgoTree can be extracted. From the ErgoTree, we can evaluate to the SigmaBoolean.

Reply:
We can evaluate ErgoTree to SigmaBoolean only in the presence of Context data. This brings us back to ReducedTransaction as a way to avoid having the Context on the wallet. While Appkit fully supports ErgoTree evaluation, other wallet app platforms may not. 

For this reason, you cannot reduce an arbitrary address to SigmaBoolean for further signing on such incapable platforms. 

But, if a dApp takes care of Address to SigmaBoolean reduction, then it can support the case you are describing and also present plain simple SigmaBoolean to the wallet for signing.

Reply:
Okay, when we want to use `SigmaBoolean`, this makes it more easy on the wallet side. To keep it simple for dApps that only want to verify p2pk addresses, we could use both ways (like there are two methods on appkit implemented).
But: is there already a serialization format for `SigmaBoolean` that is implemented in both appkit and sigma-rust?

Reply:
Yes, in Sigma it is https://github.com/ScorexFoundation/sigmastate-interpreter/blob/985ce41e2e44d82a44ac57b700c7c735f855428e/sigmastate/src/main/scala/sigmastate/Values.scala#L799
and I belive something similar exists in sigma-rust, because SigmaBoolean is part of the protocol.